File: c/src/constraints.c
#include "constraints.h"
#include <stdlib.h>
#include <string.h>

#define INITIAL_CAPACITY 8

ConstraintSet* constraint_set_create(void) {
    ConstraintSet* set = (ConstraintSet*)calloc(1, sizeof(ConstraintSet));
    if (!set) return NULL;
    set->capacity = INITIAL_CAPACITY;
    set->constraints = (Constraint*)calloc(set->capacity, sizeof(Constraint));
    return set;
}

void constraint_set_destroy(ConstraintSet* set) {
    if (!set) return;
    free(set->constraints);
    free(set);
}

void constraint_set_add(ConstraintSet* set, const Constraint* c) {
    if (!set || !c) return;
    if (set->count >= set->capacity) {
        size_t new_cap = set->capacity * 2;
        Constraint* new_arr = (Constraint*)realloc(set->constraints, new_cap * sizeof(Constraint));
        if (!new_arr) return;
        set->constraints = new_arr;
        set->capacity = new_cap;
    }
    set->constraints[set->count++] = *c;
}

void constraint_set_clear(ConstraintSet* set) {
    if (set) set->count = 0;
}

ViolationList* violation_list_create(void) {
    ViolationList* list = (ViolationList*)calloc(1, sizeof(ViolationList));
    if (!list) return NULL;
    list->capacity = INITIAL_CAPACITY;
    list->violations = (Violation*)calloc(list->capacity, sizeof(Violation));
    return list;
}

void violation_list_destroy(ViolationList* list) {
    if (!list) return;
    free(list->violations);
    free(list);
}

void violation_list_add(ViolationList* list, ValueType type, const char* message) {
    if (!list || !message) return;
    if (list->count >= list->capacity) {
        size_t new_cap = list->capacity * 2;
        Violation* new_arr = (Violation*)realloc(list->violations, new_cap * sizeof(Violation));
        if (!new_arr) return;
        list->violations = new_arr;
        list->capacity = new_cap;
    }
    list->violations[list->count].type = type;
    strncpy(list->violations[list->count].message, message, sizeof(list->violations[0].message) - 1);
    list->violations[list->count].message[sizeof(list->violations[0].message) - 1] = '\0';
    list->count++;
}

void violation_list_clear(ViolationList* list) {
    if (list) list->count = 0;
}

bool constraint_overrides_intent(const Constraint* c, ValueType type) {
    if (!c) return false;
    return c->target_type == type && (c->kind == CONSTRAINT_FIX_VALUE || c->kind == CONSTRAINT_FORBID_ASPECT);
}

void constraints_apply(const ConstraintSet* constraints, ValueStateSet* states, ViolationList* violations) {
    if (!constraints || !states || !violations) return;
    for (size_t i = 0; i < constraints->count; i++) {
        const Constraint* c = &constraints->constraints[i];
        if (c->target_type >= VALUE_TYPE_COUNT) continue;
        ValueState* vs = &states->states[c->target_type];
        switch (c->kind) {
            case CONSTRAINT_FIX_VALUE:
                *vs = value_state_present(c->fixed_val);
                break;
            case CONSTRAINT_FORBID_ASPECT:
                if (value_state_is_present(vs)) {
                    *vs = value_state_invalid(c->reason[0] ? c->reason : "forbidden");
                    violation_list_add(violations, c->target_type, "Constraint forbids this aspect");
                }
                break;
            case CONSTRAINT_LIMIT_RANGE:
                if (value_state_is_present(vs) && (vs->value < c->min_val || vs->value > c->max_val)) {
                    double clamped = vs->value < c->min_val ? c->min_val : c->max_val;
                    vs->value = clamped;
                    violation_list_add(violations, c->target_type, "Value clamped to constraint range");
                }
                break;
        }
    }
}

File: rust/src/constraints.rs
use crate::value_state::{ValueState, ValueStateSet, ValueType};
use std::vec::Vec;

#[derive(Debug, Clone)]
pub enum ConstraintKind {
    FixValue,
    ForbidAspect,
    LimitRange,
}

#[derive(Debug, Clone)]
pub struct Constraint {
    pub kind: ConstraintKind,
    pub target_type: ValueType,
    pub min_val: f64,
    pub max_val: f64,
    pub fixed_val: f64,
    pub forbid: bool,
    pub reason: String,
}

#[derive(Debug, Clone)]
pub struct ConstraintSet {
    constraints: Vec<Constraint>,
}

impl ConstraintSet {
    pub fn new() -> Self {
        Self {
            constraints: Vec::new(),
        }
    }

    pub fn add(&mut self, c: Constraint) {
        self.constraints.push(c);
    }

    pub fn clear(&mut self) {
        self.constraints.clear();
    }

    pub fn iter(&self) -> impl Iterator<Item = &Constraint> {
        self.constraints.iter()
    }
}

impl Default for ConstraintSet {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone)]
pub struct Violation {
    pub value_type: ValueType,
    pub message: String,
}

#[derive(Debug, Clone)]
pub struct ViolationList {
    violations: Vec<Violation>,
}

impl ViolationList {
    pub fn new() -> Self {
        Self {
            violations: Vec::new(),
        }
    }

    pub fn add(&mut self, value_type: ValueType, message: impl Into<String>) {
        self.violations.push(Violation {
            value_type,
            message: message.into(),
        });
    }

    pub fn clear(&mut self) {
        self.violations.clear();
    }

    pub fn is_empty(&self) -> bool {
        self.violations.is_empty()
    }

    pub fn len(&self) -> usize {
        self.violations.len()
    }

    pub fn iter(&self) -> impl Iterator<Item = &Violation> {
        self.violations.iter()
    }
}

impl Default for ViolationList {
    fn default() -> Self {
        Self::new()
    }
}

pub fn constraint_overrides_intent(c: &Constraint, value_type: ValueType) -> bool {
    c.target_type == value_type
        && matches!(c.kind, ConstraintKind::FixValue | ConstraintKind::ForbidAspect)
}

pub fn constraints_apply(
    constraints: &ConstraintSet,
    states: &mut ValueStateSet,
    violations: &mut ViolationList,
) {
    for c in constraints.iter() {
        if c.target_type != ValueType::TempoBias
            && c.target_type != ValueType::HarmonicTension
            && c.target_type != ValueType::RhythmicDensity
            && c.target_type != ValueType::DynamicRange
        {
            continue;
        }
        let state = states.get(c.target_type);
        match &c.kind {
            ConstraintKind::FixValue => {
                states.set(c.target_type, ValueState::present(c.fixed_val));
            }
            ConstraintKind::ForbidAspect => {
                if state.is_present() {
                    states.set(
                        c.target_type,
                        ValueState::invalid(if c.reason.is_empty() {
                            "forbidden"
                        } else {
                            &c.reason
                        }),
                    );
                    violations.add(c.target_type, "Constraint forbids this aspect");
                }
            }
            ConstraintKind::LimitRange => {
                if let ValueState::Present(v) = state {
                    if *v < c.min_val || *v > c.max_val {
                        let clamped = if *v < c.min_val {
                            c.min_val
                        } else {
                            c.max_val
                        };
                        states.set(c.target_type, ValueState::present(clamped));
                        violations.add(c.target_type, "Value clamped to constraint range");
                    }
                }
            }
        }
    }
}
